### PR TITLE
feat: xblock verification event handler

### PR DESCRIPTION
**Pull request by navinkarkera**
**Description**

Adds handler to take `XBLOCK_SKILL_VERIFIED` signal from openedx-events implemented in https://github.com/open-craft/openedx-events/pull/1 and updates counts of related xblock skills.

**Related MRs**

- https://github.com/openedx/openedx-events/pull/165
- https://github.com/open-craft/xblock-skill-tagging/pull/1

**Test instructions**

- Checkout this branch. (No need for devstack, this can be tested independently)
- Create a virtual environment with python 3.8 and activate it.
- Run `make requirements` and `python manage.py migrate`.
- Open django shell using `python manage.py shell`.
- Run below piece of code
```python
# fill dummy data
import factory
import factory.fuzzy
from taxonomy import models
from test_utils.factories import *
XBlockSkillsFactory.create_batch(5)
SkillFactory.create_batch(20)
XBlockSkillDataFactory.create_batch(20, skill=factory.Iterator(models.Skill.objects.all()), xblock=factory.Iterator(models.XBlockSkills.objects.all()), verified=factory.fuzzy.FuzzyChoice([True, False]))
# get first xblock
xblock=XBlockSkills.objects.first()
skills = XBlockSkillData.objects.filter(xblock=xblock).values("id", "verified_count", "ignored_count")
# check that all counts are 0
print(skills)
# fire event
from openedx_events.learning.data import XBlockSkillVerificationData
from openedx_events.learning.signals import XBLOCK_SKILL_VERIFIED
XBLOCK_SKILL_VERIFIED.send_event(
    xblock_info=XBlockSkillVerificationData(
        usage_key=xblock.usage_key,
        verified_skills=[skills[0]["id"], skills[1]["id"]],
        ignored_skills=[skills[2]["id"], skills[3]["id"]],
    )
)
# verify that counts are increament
skills = XBlockSkillData.objects.filter(xblock=xblock).values("id", "verified_count", "ignored_count")
print(skills)
```

**Merge checklist:**
- [x] Any new requirements are in the right place (do **not** manually modify the `requirements/*.txt` files)
    - `make upgrade && make requirements` have been run to regenerate requirements
- [x] `./manage.py makemigrations` has been run
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
- [x] [Version](https://github.com/edx/taxonomy-connector/blob/master/taxonomy/__init__.py) bumped
- [x] [Changelog](https://github.com/edx/taxonomy-connector/blob/master/CHANGELOG.rst) record added

**Post merge:**
- [ ] Tag pushed and a new [version](https://github.com/edx/taxonomy-connector/releases) released
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
- [ ] After versioned build finishes in [GitHub Actions](https://github.com/edx/taxonomy-connector/actions), verify version has been pushed to [PyPI](https://pypi.org/project/taxonomy-connector/)
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
- [ ] PR created in [course-discovery](https://github.com/edx/course-discovery) to upgrade dependencies (including taxonomy-connector)
    - This **must** be done after the version is visible in PyPi as `make upgrade` in course-discovery will look for the latest version in PyPi.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @navinkarkera!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

File: taxonomy/signals/handlers.py [hunk 1]
```diff
@@ -85,3 +88,20 @@ def handle_xblock_published(**kwargs):  # pylint: disable=unused-argument
...
+@receiver(XBLOCK_SKILL_VERIFIED)
+def handle_xblock_verified(**kwargs):
+    """
+    Handle signal and trigger task to update xblock skills verified and ignored count.
+    """
+    LOGGER.info('[TAXONOMY] XBLOCK_SKILL_VERIFIED signal received')
+    xblock_data = kwargs.get('xblock_info', None)
+    if not xblock_data or not isinstance(xblock_data, XBlockSkillVerificationData):
+        LOGGER.error('[TAXONOMY] Received null or incorrect data from XBLOCK_SKILL_VERIFIED.')
+        return
```
Comment (pkulkark) [inline taxonomy/signals/handlers.py:102]: Does it make sense to add a check to see if verified_skills or ignored_skills is empty?

File: taxonomy/tasks.py [hunk 2]
```diff
@@ -90,3 +90,17 @@ def duplicate_xblock_skills(source_xblock_uuid, xblock_uuid):
...
+    """
+    Task to update xblock skills verification counts.
+
+    Arguments:
+        xblock_uuid (str): uuid of xblock.
+        verified_skills (List[int]): list of verified skill ids.
+        ignored_skills (List[int]): list of ignored skill ids.
+    """
+    LOGGER.info('[TAXONOMY] update_xblock_skills_verification_counts task triggered')
+    utils.update_xblock_skills_verification_counts(xblock_uuid, verified_skills, ignored_skills)
```
Comment (pkulkark) [inline taxonomy/tasks.py:106]:
```suggestion
    utils.update_xblock_skills_verification_counts(xblock_uuid, verified_skills, ignored_skills)
    LOGGER.info('[TAXONOMY] update_xblock_skills_verification_counts task completed')
```

File: taxonomy/utils.py [hunk 3]
```diff
@@ -662,3 +663,31 @@ def duplicate_xblock_skills(source_xblock_uuid, xblock_uuid):
...
+    """
+    Update xblock skill verified and ignored count.
+
+    Arguments:
+        usage_key (str): usage_key of xblock.
+        verified_skills (List[int]): list of verified skill ids.
+        ignored_skills (List[int]): list of ignored skill ids.
+    """
+    xblock_model, identifier = get_product_skill_model_and_identifier(ProductTypes.XBlock)
+    xblock = xblock_model.objects.filter(**{identifier: usage_key}).first()
```
Comment (pkulkark) [inline taxonomy/utils.py:678]: Why not just use `objects.get` instead of `objects.filter` if we're expecting a single xblock?

Comment (navinkarkera) [reply taxonomy/utils.py:678]: @pkulkark Get will raise an error if xblock is not found. This way we can check and log an error and return (missed to add return statement, updated!).

Comment (pkulkark) [reply taxonomy/utils.py:678]: @navinkarkera We could put it in a try-catch block and log the required error in the catch block and return. I'm not completely opposed to using `filter` but it just feels a bit unconventional.

File: tests/test_signals.py [hunk 4]
```diff
@@ -178,3 +181,95 @@ def test_empty_event_data_format_skips_processing(self):
...
+                    ignored_skills=[],
+                )
+            )
+            messages = [record.msg for record in log_capture.records]
+            self.assertEqual(len(log_capture.records), 6)
+            self.assertEqual(
+                [
+                    f'[TAXONOMY] XBLOCK_SKILL_VERIFIED signal received',
+                    '[TAXONOMY] update_xblock_skills_verification_counts task triggered',
+                    f'[TAXONOMY] XBlock with usage_key: {self.xblock.key} not found!',
```
Comment (pkulkark) [inline tests/test_signals.py:248]: Shouldn't this contain the task completed log as well, if we're returning after logging error message?

File: taxonomy/utils.py (hunk 3 above)
Comment (navinkarkera) [reply taxonomy/utils.py:678]: @pkulkark The filter option avoids try/catch boilerplate, but I agree that using `get` makes more sense here as we want only one object. Updated :+1:

File: tests/test_signals.py (hunk 4 above)
Comment (navinkarkera) [reply tests/test_signals.py:248]: Yes, it was not checking for it, see next line: `messages[:-3]`. Updated it to check for it, thanks.

**Review by pkulkark — APPROVED** [review]
@navinkarkera  LGTM :+1:

- [x] I tested this: I verified the xblock verification event handler works as described in the testing instructions
- [x] I read through the code
- [ ] ~I checked for accessibility issues~ N/A
- [ ] ~Includes documentation~ N/A

**Comment by mphilbrick211** [comment]
Hi @navinkarkera - is this ready for review / merge?

**Comment by navinkarkera** [comment]
@mphilbrick211 Yes.
cc: @sameenfatima78

File: taxonomy/utils.py [hunk 5]
```diff
@@ -662,3 +664,33 @@ def duplicate_xblock_skills(source_xblock_uuid, xblock_uuid):
...
+        xblock=xblock,
+        skill_id__in=verified_skills,
+        is_blacklisted=False,
+    ).update(verified_count=F('verified_count') + 1)
+    # update ignored_count
+    xblock_data_model.objects.filter(
+        xblock=xblock,
+        skill_id__in=ignored_skills,
+        is_blacklisted=False,
+    ).update(ignored_count=F('ignored_count') + 1)
```
Comment (sameenfatima78) [inline taxonomy/utils.py:696]: Question: Can you explain why are we adding +1 to both counts here?

**Review by sameenfatima78 — COMMENTED** [review]
One question plus version needs to be upated. Also, the point that Mike raised in our last call, will that scenario be handled in this PR?

File: taxonomy/utils.py (hunk 5 above)
Comment (navinkarkera) [reply taxonomy/utils.py:696]: @sameenfatima78  `ignored_skills` contains ids of skills which were ignored and `verified_skills` contain ids of verified skills. So we just take them and increment their counts. If it is empty nothing will be updated.

**Comment by navinkarkera** [comment]
@sameenfatima78 Updated MR. We need to wait till https://github.com/openedx/openedx-events/pull/165 is merged.

File: taxonomy/utils.py (hunk 5 above)
Comment (sameenfatima78) [reply taxonomy/utils.py:696]: It's a bit confusing that on `XBLOCK_SKILL_VERIFIED` signal handler we are updating the `ignored_count`. I'm assuming this signal will be called when an `xblock` is verified, is this correct?

**Review by sameenfatima78 — COMMENTED** [review]
Added a comment.

File: taxonomy/utils.py (hunk 5 above)
Comment (navinkarkera) [reply taxonomy/utils.py:696]: [XBLOCK_SKILL_VERIFIED](https://github.com/openedx/openedx-events/pull/165/files#diff-5cf3a0b26961626ce777acf38a3cd803cb0ab982f8f35a35d078cb9c62438fa3R223-R235) is sending both verified and ignored skill ids as there is only one action to submit this information.

SS

Comment (sameenfatima78) [reply taxonomy/utils.py:696]: Ah okay, got it. Thanks for clarifying.

**Review by sameenfatima78 — APPROVED** [review]

**Comment by mphilbrick211** [comment]
Hi @navinkarkera - is this pull request ready for merge? If so, would you mind addressing the branch date error?

**Comment by navinkarkera** [comment]
@sameenfatima78 https://github.com/openedx/openedx-events/pull/165 was merged and I have updated the dependencies now. This MR is ready to be merged.

cc: @mphilbrick211

**Comment by openedx-webhooks** [comment]
@navinkarkera 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/taxonomy-connector&entry.867055334=https://github.com/openedx/taxonomy-connector/pull/135&entry.1484655318=https://github.com/navinkarkera&entry.752974735=2022-12-27+06:45&entry.1917517419=2023-02-20+20:29&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**